cxx-squid/src/main/java/org/sonar/cxx/preprocessor/CxxPreprocessor.java

```diff
@@ -478,54 +535,32 @@ public void init() {
           // Treat all global defines as high prio
           getMacros().setHighPrio(true);
 
-          // parse the configured defines and store into the macro library
-          for (String define : conf.getDefines()) {
-            LOG.debug("parsing external macro to unit: '{}'", define);
-            if (!"".equals(define)) {
-              Macro macro = parseMacroDefinition("#define " + define);
-              if (macro != null) {
-                LOG.debug("storing external macro to unit: '{}'", macro);
-                getMacros().put(macro.name, macro);
-              }
-            }
-          }
-
-          // set standard macros
-          // using smaller set of defines as rest is provides by compilation unit settings
-          getMacros().putAll(Macro.UNIT_MACROS);
-
-          // parse the configured force includes and store into the macro library
-          for (String include : conf.getForceIncludeFiles()) {
-            LOG.debug("parsing force include to unit: '{}'", include);
-            if (!"".equals(include)) {
-              // TODO -> this needs to come from language
-              parseIncludeLine("#include \"" + include + "\"", "sonar.cxx.forceIncludes", conf.getEncoding());
-            }
-          }
+          // add macros which are forced for each compilation unit
+          getMacros().putAll(forcedUnitMacros);
 
           // rest of defines comes from compilation unit settings
-          for (Map.Entry<String, String> entry : compilationUnitSettings.getDefines().entrySet()) {
-            final String name = entry.getKey();
-            final String body = entry.getValue();
-            getMacros().put(name, new Macro(name, body));
+          if (useGlobalCUSettings) {
+            getMacros().putAll(globalCUDefines);
+          } else {
+            getMacros().putAll(parseMacroDefinitions(compilationUnitSettings.getDefines()));
           }
         } finally {
           getMacros().setHighPrio(false);
         }
 
-        if (getMacro(CPLUSPLUS) == null) {
+        if (getMacro(Macro.CPLUSPLUS) == null) {
           //Create macros to replace C++ keywords when parsing C files
           getMacros().putAll(Macro.COMPATIBILITY_MACROS);
         }
       } else {
         // Use global settings
         LOG.debug("global settings for: '{}'", currentContextFile);
-        if (isCFile(currentContextFile.getAbsolutePath())) {
+        if (isCFile(currentContextFile.getName())) {
           //Create macros to replace C++ keywords when parsing C files
           getMacros().putAll(Macro.COMPATIBILITY_MACROS);
-          fixedMacros.disable(CPLUSPLUS);
+          getMacros().disable(Macro.CPLUSPLUS);
```

Comment:
think there are more macros different between C and C++?

Reply:
the map `Macro.STANDARD_MACROS_IMPL` contains only two C++ specific macros: 

* __cplusplus
* __has_include (since C++17)

so theoretically we have to disable `__has_include` to. BUT

1. I tested GCC in the ANSI-C mode and it still provides this define (https://coliru.stacked-crooked.com/a/30a53f2c2a9a323d)
2. Our implementation of `__has_include` is hard-wired in the grammar and the expansion of `__has_include()` is not restricted to C++. So even if `#ifdef __has_include` evaluates to `0`, we still can't prevent `__has_include(...)` from evaluation.
3. This C vs C++ was not scope of my change, so please let's consider to change it in other PR